### PR TITLE
Account directory page demo

### PR DESCRIPTION
**Pull request by spiresheep**
This pull request has the discussed API changes. The test related to AccountDirectoryPage has been changed to still be functional.

This also includes a fix for the transitions when a card is open/closed.

### CODE REVIEW

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_controller.tsx [hunk 1]
```diff
@@ -1,6 +1,9 @@
+import * as Nexus from 'nexus';
 import * as React from 'react';
 import { DisplaySize } from '../..';
 import { LoadingPage } from '..';
+import { AccountDirectoryModel } from '.';
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_controller.tsx:5]: Going forward we no longer allow ```import from '.'``` You can import from ```..``` or ```../..```, it's just ```.``` that is not allowed due to causing issues with circular dependencies.

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_controller.tsx [hunk 2]
```diff
@@ -23,6 +26,8 @@ export class AccountDirectoryController extends
   }
 
   public render(): JSX.Element {
+    if(!this.state.isLoaded) {
     return <LoadingPage/>;
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_controller.tsx:30]: Indentation.

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_page.tsx [hunk 3]
```diff
@@ -13,43 +12,42 @@ interface Properties {
...
+  openedGroups: Beam.Map<Beam.DirectoryEntry, AccountEntry[]>
+
+  /** The filter used to return a subset of groups. */
+  filter : string;
+
+  /** The accounts that match the current filter. */
+  filteredGroups: Beam.Map<Beam.DirectoryEntry, AccountEntry[]>
+
+  /** Called when the filter value changes. */
+  onFilterChange?:(filter: string) => void;
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_page.tsx:31]: Space after ```:```.

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_page.tsx [hunk 4]
```diff
@@ -13,43 +12,42 @@ interface Properties {
...
+
+  /** Called when a card is clicked on. */
+  onCardClick?:(group: Beam.DirectoryEntry) => void;
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_page.tsx:34]: Space after ```:```

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_page.tsx [hunk 5]
```diff
@@ -13,43 +12,42 @@ interface Properties {
...
-  filter: string;
-  accounts: Beam.Map<Beam.DirectoryEntry, AccountEntry[]>;
-  openedGroups: Beam.Set<Beam.DirectoryEntry>;
-}
 
 /** Displays a directory of accounts. */
-export class AccountDirectoryPage extends React.Component<Properties, State> {
+export class AccountDirectoryPage extends React.Component<Properties> {
   public static readonly defaultProps = {
     onNewGroupClick: () => {},
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_page.tsx:47]: Optional props should have default values here.

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_page.tsx [hunk 6]
```diff
@@ -128,58 +149,14 @@ export class AccountDirectoryPage extends React.Component<Properties, State> {
...
-                  onClick={this.props.onNewAccountClick}>
-                New Account
-              </button>
-              <div style={AccountDirectoryPage.STYLE.spacing}/>
-              <button onClick={this.props.onNewGroupClick}
-                  className={css(buttonStyle)}>
-                New Group
-              </button>
+            <div style={AccountDirectoryPage.STYLE.spacing}/>
+            <div id='group_cards'>
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_page.tsx:154]: Remove the ```id```.

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_page.tsx [hunk 7]
```diff
@@ -101,23 +99,46 @@ export class AccountDirectoryPage extends React.Component<Properties, State> {
...
     return (
-      <div style={AccountDirectoryPage.STYLE.page}>
-        <div style={contentWidth}>
-          <div id='header' style={headerBoxStyle}>
-            <div style={verticalButtonVisibility}>
-              <div style={buttonBoxStyle}>
+      <PageWrapper>
+        <div style={AccountDirectoryPage.STYLE.page}>
+          <div style={contentWidth}>
+            <div id='header' style={headerBoxStyle}>
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_page.tsx:123]: Remove ```id```.

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_entry_row.tsx [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+      if(this.props.account) {
+        return AccountEntryRow.DYNAMIC_STYLE.entry;
+      } else {
+        return null;
+      }
+    })();
+    const animation_style = (() => {
+      if(!this.props.filter) {
+        return AccountEntryRow.animationStyle;
+      }else {
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_entry_row.tsx:76]: Space after ```}```

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/group_card.tsx [hunk 9]
```diff
@@ -27,13 +27,15 @@ interface Properties {
   onDropDownClick?: () => void;
 
   /** Called when a directory is clicked on.
-   * @param entry - A accounty or directory entry.
+   * @param entry - A account or directory entry.
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/pages/account_directory_page/group_card.tsx:30]: ```An account```.

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/group_card.tsx [hunk 10]
```diff
@@ -111,99 +105,77 @@ export class GroupCard extends React.Component<Properties, State> {
...
-            timeout={GroupCard.TIMEOUTS}>
-          {(state) => (
-            <div>
-              {lineWhenOpen}
-              <div style={GroupCard.STYLE.entryListWrapper}>
-                {accounts}
-              </div>
-              <div style={
-                    (GroupCard.bottomPaddingAnimationStyle as any)[state]}>
+        <div>
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/pages/account_directory_page/group_card.tsx:150]: What happens if you remove this ```div```?

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/group_card.tsx [hunk 11]
```diff
@@ -111,99 +105,77 @@ export class GroupCard extends React.Component<Properties, State> {
...
+          </Transition>
+        </div>
       </VBoxLayout>);
   }
 
+  public static getDerivedStateFromProps(props: Properties, state: State) {
+    if(!props.isOpen && !state.isOpen && props.accounts.length !==
+        state.localAccounts.length) {
+      return {localAccounts: props.accounts};
+    }else if(props.isOpen && !state.isOpen) {
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/pages/account_directory_page/group_card.tsx:170]: Space.

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/index.ts [hunk 12]
```diff
@@ -2,6 +2,7 @@ export * from './account_directory_controller';
 export * from './account_directory_model';
 export * from './account_directory_page';
 export * from './account_entry';
+export * from './account_entry_row';
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/pages/account_directory_page/index.ts:5]: Probably don't need to export this, this is an internal class.

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/group_card.tsx (hunk 10 above)
Comment (spiresheep) [reply Applications/WebPortal/WebApp/library/source/pages/account_directory_page/group_card.tsx:150]: Nothing, so it shall be gone.

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_page.tsx [hunk 13]
```diff
@@ -151,7 +153,7 @@ export class AccountDirectoryPage extends React.Component<Properties> {
               </div>
             </div>
             <div style={AccountDirectoryPage.STYLE.spacing}/>
-            <div id='group_cards'>
+            <div>
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/pages/account_directory_page/account_directory_page.tsx:156]: Is this ```<div>``` needed?

File: Applications/WebPortal/WebApp/library/source/pages/account_directory_page/group_card.tsx [hunk 14]
```diff
@@ -147,7 +147,7 @@ export class GroupCard extends React.Component<Properties, State> {
             {this.props.group.name}
           </div>
         </div>
-        <div>
+
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/pages/account_directory_page/group_card.tsx:150]: Remove blank line, fix indentation.

**Pull request merged**